Authorizer support in Websocket

computerpunc opened the pull request:
This PR contains support for authorizer in WebSocket by running an auth function before `$connect`. 
Currently supports the HTTP header `Auth` only (i.e., no support for identitySource yet).

https://serverless.com/framework/docs/providers/aws/events/websocket#using-authorizers

Also fixes #753.

Main change with this PR is using a modified version of `hapi-plugin-websocket` to overcome its limitations in supporting auth prior to `$connect`.

To Do:

- [x] Lint
- [x] Jest
- [x] Documentation
- [x] identitySource

dnalborczyk commented:
@computerpunc awesome!! don't worry too much about the conflicting files. just moving some stuff around.I can help resolving the conflicts.

btw, I simplified parseQuery https://github.com/dherault/serverless-offline/commit/c83980e32edbd6d356add59440033258b4e62c9b a bit, incl. some tests, and moved it to https://github.com/dherault/serverless-offline/commit/8ca29f8baf252a4b705e091c30adb0bdd366c76a

computerpunc commented:
> btw, I simplified parseQuery [c83980e](https://github.com/dherault/serverless-offline/commit/c83980e32edbd6d356add59440033258b4e62c9b) a bit, incl. some tests, and moved it to [8ca29f8](https://github.com/dherault/serverless-offline/commit/8ca29f8baf252a4b705e091c30adb0bdd366c76a) . when you have a minute, could you check if it covers what `parseQuery` intended?
> 

Since moving from `req.url` in `hapi-plugin-websocket`'s `connect` to `request.url` in `hapi`'s `handler`. The `url` is not a string but a URL object. I was going to remove `parseQuery` altogether because you can easily get the params from `url.seachParams`. Just not in the right object format and your code with object-from-entries solves it.
After seeing all the moving around, I decided to make a PR as soon as possible to postpone any changes to widen the gap between master and this branch. So I didn't have time to handle it.

dnalborczyk commented:
ah, cool, thank you!
> So I didn't have time to handle it.

 no worries.

computerpunc commented:
> don't worry too much about the conflicting files. just moving some stuff around.I can help resolving the conflicts.

This branch has 3 conflicts with master in `ApiGatewayWebSocket.js` . I can merge and resolve without a problem.

dnalborczyk commented:
> The url is not a string but a URL object. 

would it help to "overload" parseQueryParameter to take an `URL` object as well?

dnalborczyk commented:
@computerpunc I changed it to just take a [SearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) object as parameter. Also added multiValueQueryStringParameters + tests. Haven't switched both with the HTTP implementation yet.

computerpunc commented:
@dnalborczyk I see that you make all the changes in master. In comparison to this PR branch, it is out of date regarding WS. Can you make your changes directly in this PR (or a least on code that is derived from this PR)?

computerpunc commented:
Guys,

I finished all the pending issues or my list. Anything else? Merge?

dherault commented on src/ApiGatewayWebSocket.js:
```diff
@@ -2,15 +2,13 @@
 const h2o2 = require('@hapi/h2o2');
-const hapiPluginWebsocket = require('hapi-plugin-websocket');
+const boom = require('@hapi/boom');
+const hapiPluginWebsocket = require('./hapi-plugin-websocket');
```
Why is the Hapi plugin in a separate file?

dnalborczyk commented:
@computerpunc Thank you so much for putting in the time and effort!!!

regarding the conflicts: I removed the `package-lock.json` from all manual tests, as it is linking the project root and "locking" the dependencies makes a bit less sense there. so nothing to merge, just remove.

> In comparison to this PR branch, it is out of date regarding WS. Can you make your changes directly in this PR

I try to keep it to a minimum or ideally not affecting anything `websocket` related for now at all. just trying to clean up a bit as this project has grown quite a bit over time, and it's really hard to add new features or even fix bugs. it seems there's quite a bit of redundancies in the code.

> I finished all the pending issues or my list. Anything else?

I might need a bit of time to look over this.

computerpunc commented:
Guys,

This is a bit unexpected that while waiting for the review of this PR, I improved the **manual** testing of WebSocket to the level that it is no longer needed to run **manually** but can run **automatically**. 

Running `npm run test-e2e` will spawn locally test severs, run all test suites against the servers and then clean up.

How about we step up WebSocket testing and add them to the CI? 
At first, failing WebSocket tests won't fail the whole CI process, it will be just optional, and then as we get more confident, we can move the tests from `src/__tests__/manual` to `src/__tests__/e2e` and set it as mandatory for the CI.

yareyaredesuyo commented:
Is there anyone reviewing this issue?

Looks like penging man.

computerpunc commented:
@dnalborczyk @dherault 

Guys,
What is the reason this experimental WebSocket PR that is backed by 31 100% AWS compatible tests and includes a feature, an automatic test suite that completes in less than 5 seconds and numerous bug fixes, is waiting so long to be merged?

Each time someone touches ApiGatewayWebSocket.js or ServerlessOfline.js, there's a merge conflict that I have to manually resolve. Very tedious :(

BTW the automatic WebSocket tests are now all red because it seems that the plugin doesn't pickup configuration info such as:
```
custom:
  serverless-offline:
    port: 3004
    websocketPort: 3005
```

dnalborczyk commented:
hey @computerpunc  I have been on vacation and literally just came back.

I wanna spend some time the next couple days to have a closer look at your PR.

> BTW the automatic WebSocket tests are now all red because it seems that the plugin doesn't pickup configuration info such as:

are you using the following setup to run the tests: https://github.com/dherault/serverless-offline/tree/master/src/__tests__/manual/websocket ?

computerpunc commented:
OK so I'll stop trying to keep this PR in non conflicting state. Whenever you're ready, just signal and I'll resolve the conflicts manually.

> are you using the following setup to run the tests: https://github.com/dherault/serverless-offline/tree/master/src/__tests__/manual/websocket ?

`master` branch regarding WebSocket is really ancient tech nowadays :) I'm using https://github.com/computerpunc/serverless-offline/tree/master/src/__tests__/manual/websocket
If you want to have a go with testing read the README file first.

As to the problem with picking up the configuration, just add the following to any existing `serverless.yml` files:
```
custom:
  serverless-offline:
    port: 7000
```
And when you run `sls offline` the following line appears `Serverless: Offline [HTTP] listening on http://localhost:3000` instead of `Serverless: Offline [HTTP] listening on http://localhost:7000`.

Because it's not picking the configuration it's using default values (port 3000 in this case).

dnalborczyk commented:
> As to the problem with picking up the configuration, just add the following to any existing `serverless.yml` files:
> 
> ```
> custom:
>   serverless-offline:
>     port: 7000
> ```
> 
> And when you run `sls offline` the following line appears `Serverless: Offline [HTTP] listening on http://localhost:3000` instead of `Serverless: Offline [HTTP] listening on http://localhost:7000`.
> 
> Because it's not picking the configuration it's using default values (port 3000 in this case).

Just checked `master`, it doesn't seem to have this problem.

computerpunc commented:
> Just checked `master`, it doesn't seem to have this problem.

It still does for me.
Last commit that doesn't experience this problem `980ea0bbe29450b4c66676c175b795f3e4a5da59`.
First commit to experience this problem `91380030896b8f7f8fc9ea6ec34f08f24ffab949`

dnalborczyk commented:
> It still does for me.

did you check `master`, or your branch?

computerpunc commented:
> > It still does for me.
>
> did you check `master`, or your branch?

`master`

dnalborczyk commented:
I'm not able to repro. it seems to work just fine. Could you check that you are not overwriting the `custom parameter` with a `cli option` (which takes precedence)?

Are you able to pinpoint the problem so I could have a look?

computerpunc commented:
> Are you able to pinpoint the problem so I could have a look?

Now each entry in `this.commands.offline.options` has a default value. E.g., `this.commands.offline.options.port.default` is 3000.

What doesn't it mean? 

It means that if you don't supply a value on the `cli`, `serverless` will supply a default value as if you supplied it on the `cli`.
And since `cli` has precedence, the default value can be the only value (unless another value is supplied on the `cli`).

In practice, the default values are injected to `this.options` by `serverless` before calling `offline:start:`.

dnalborczyk commented:
> > Are you able to pinpoint the problem so I could have a look?
> 
> Now each entry in `this.commands.offline.options` has a default value. E.g., `this.commands.offline.options.port.default` is 3000.

yes, I moved (consolidated) the `defaults` into the `commands object`.

> It means that if you don't supply a value on the `cli`, `serverless` will supply a default value as if you supplied it on the `cli`.

That's what I thought it would do as well, **but it does not** (not sure if this is a bug, I probably have to open an issue @ serverless).

computerpunc commented:
> * could you check the `serverless` version?
```
$ sls --version
1.48.4 (Enterprise Plugin: 1.3.2, Platform SDK: 2.1.0)
```
> * make sure you are not setting (overwriting) the cli options, if you are interested in the custom options

running plain vanilla:
```
sls offline
```
> * `master` is latest

Yes.

dnalborczyk commented:
thanks for checking!

gonna create a standalone repo, if it still doesn't repro, I'm gonna push it to github.

computerpunc commented:
In `ServerlessOffline.js`, can you check the following?
1. The value of `options` in
```
module.exports = class ServerlessOffline {
  constructor(serverless, options) {
```
2. The value of `this.options` **before** the code:
```
this.options = {
      ...defaultOptions,
      ...defaultOptionsTEMP, // TODO FIXME, see above
      ...customOptions,
      ...this.options,
    };
```

dnalborczyk commented:
> The value of this.options before the code

I did, empty object `{}`.

dnalborczyk commented:
I'm wondering if the problem might be the ES6 module loader I'm using, since the property's name is `default`. We'll see, gonna create a small repo ...

dnalborczyk commented:
ah, nevermind, found it. difference is `sls offline` vs. `sls offline start`.

kaoskeya commented:
> Currently supports the HTTP header `Auth` only (i.e., no support for identitySource yet).

Hey @computerpunc , I got the `identitySource` to work (headers and query parameter). I am unable to run the manual tests though, if you could point me on how to run the tests, I can write that out, implement the support for identitySource and send a PR.

Also, thanks for all the work you've put in.

computerpunc commented:
> > Currently supports the HTTP header `Auth` only (i.e., no support for identitySource yet).
> 
> Hey @computerpunc , I got the `identitySource` to work (headers and query parameters). I am unable to run the manual tests though, if you could point me on how to run the tests, I can write that out, implement the support for identitySource and send a PR.
> 

@kaoskeya When this PR started `identitySource` didn't work but it was on the to do list. Now, it works with support for headers but not for query parameters.

I presume that you made changes based on `master` and not based on this PR. 'master` is old tech compared to this PR. No real WebSocket development should really be made before the merge.

I suggest you wait until this PR is merged and then add `identitySource` query parameters support.

You can take a look at this PR for now at https://github.com/computerpunc/serverless-offline/tree/websocket-fixes-authorizer

If you want to run tests read the README file first at https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/__tests__/manual/websocket/README.md

kaoskeya commented:
> > > Currently supports the HTTP header `Auth` only (i.e., no support for identitySource yet).
> > 
> > 
> > Hey @computerpunc , I got the `identitySource` to work (headers and query parameters). I am unable to run the manual tests though, if you could point me on how to run the tests, I can write that out, implement the support for identitySource and send a PR.
> 
> @kaoskeya When this PR started `identitySource` didn't work but it was on the to do list. Now, it works with support for headers but not for query parameters.
> 
> I presume that you made changes based on `master` and not based on this PR. 'master` is old tech compared to this PR. **No additional real WebSocket development should really be made before the merge**.
> 
> I suggest you wait until this PR is merged and then add `identitySource` query parameters support.
> 
> You can take a look at this PR for now at https://github.com/computerpunc/serverless-offline/tree/websocket-fixes-authorizer

I actually cloned your repo, pulled in the `websocket-fixes-authorizer` branch and got the headers and query params to work by developing in another branch. I'll wait until this PR gets merged into master and then get my local copy up to date and send in a PR.

> If you want to run tests read the README file first at https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/__tests__/manual/websocket/README.md

Yup, I did follow the instructions here, I'll try again and should be able to get it to work now that I know for sure I'm looking at the correct file.

Thank you.

computerpunc commented:
> Yup, I did follow the instructions here, I'll try again and should be able to get it to work now that I know for sure I'm looking at the correct file.

Do let me know if you managed to run the tests. If not, I'll try to help and improve the README file if needed.

yareyaredesuyo commented:
Whats going on this PR? pending?

dnalborczyk commented:
hey @computerpunc 

I'm gonna take a break later today from shuffling stuff around. It's just been very hard to fix bugs and extend functionality, without a major restructuring and cleaning up -- and there's still plenty to do. Anyhow, I don't want to hold up this PR much longer, so we can finally go over the code and pull it in.

I'll let you know later today.

Thanks again for all the work you have done and your patience!

dnalborczyk commented:
> Whats going on this PR? pending?

@yareyaredesuyo 

yes, it's pending. Feel free to give the PR a try. Feedback would be much appreciated!

computerpunc commented:
> I'll let you know later today.

@dnalborczyk 
Waiting for your reply. Just to see that we are on the same page: I'm waiting for a green light, then I'll resolve merge conflicts and then you can merge this PR to `master`. Correct?

dnalborczyk commented:
> Waiting for your reply. Just to see that we are on the same page: I'm waiting for a green light, then I'll resolve merge conflicts and then you can merge this PR to `master`. Correct?

yeah, that's the plan. you can go ahead and resolve the conflicts.

computerpunc commented:
@dnalborczyk 
PR is ready for merge.

dnalborczyk commented:
@computerpunc 

couple observations:

- merges from 'master' seem to be borked. not quite sure what to do about that. maybe we can revert the merges somehow, and redo them?

- it definitely needs some formatting, e.g. look at this line as an example: https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/ServerlessOffline.js#L379 

- linting doesn't seem to work, that's probably on me though, I'll look into that.

- I tried to follow your test instructions https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/__tests__/manual/websocket/README.md   but didn't have much luck.  I'm also not quite sure what needs to be run in which order.

Before you fix anything, the most important thing really is a `clean merge`. I'll have to do some research on how to do about that. maybe it's not so much of a problem either, depending on what was changed/merged badly. we could just re-apply the changes by hand.

I'm thinking of pulling in the PR into a branch, same as we did with `websockets`, if that's alright with you?

computerpunc commented:
> * I tried to follow your test instructions https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/__tests__/manual/websocket/README.md   but didn't have much luck.  I'm also not quite sure what needs to be run in which order.

Having stable code that is backed by tests is the most important part. Here's what it should look like when running `nom run test-e2e`:
```
> manual_test@0.0.0 test-e2e /Users/ram/Projects/serverless/serverless-offline/src/__tests__/manual/websocket
> ./scripts/test-e2e.sh

Starting test servers ...

> manual_test@0.0.0 deploy-offline-all /Users/ram/Projects/serverless/serverless-offline/src/__tests__/manual/websocket
> node scripts/deploy-offline-all.js

Spawning offline test server 'main' [PID=4712] as a separate process ...
Spawning offline test server 'authorizer' [PID=4713] as a separate process ...
Spawning offline test server 'RouteSelection' [PID=4714] as a separate process ...
Servers are running ...

> manual_test@0.0.0 test-all /Users/ram/Projects/serverless/serverless-offline/src/__tests__/manual/websocket
> mocha test/*



  serverless
    with RouteSelection [WebSocket] support
      ✓ should call action 'echo' handler located at service.do (88ms)

  serverless
    with Authorizer [WebSocket] support
      ✓ should open a WebSocket with good auth (137ms)
      ✓ should not open a WebSocket with no auth
      ✓ should not open a WebSocket with bad auth
      ✓ should not open a WebSocket with auth function throwing exception (53ms)
      ✓ should get 401 when trying to open WebSocket with no auth
      ✓ should get 403 when trying to open WebSocket with incorrect auth
      ✓ should receive correct call info (105ms)

  serverless
    with WebSocket support
      ✓ should request to upgade to WebSocket when receving an HTTP request
      ✓ should open a WebSocket (74ms)
      ✓ should receive client connection info (48ms)
      ✓ should call default handler when no such action exists (39ms)
      ✓ should call default handler when no action provided (57ms)
      ✓ should send & receive data (107ms)
      ✓ should respond when having an internal server error (49ms)
      ✓ should get error when handler does not respond (40ms)
      ✓ should not open a connection when connect function returns an error
      ✓ should get the error when trying to open WebSocket and connect function returns an error
      ✓ should not open a connection when connect function throwing exception
      ✓ should get 502 when trying to open WebSocket and having an exeption in connect function
      ✓ should not open a connection when connect function not answer
      ✓ should get 502 when trying to open WebSocket and connect function not answer
      ✓ should respond via callback
      ✓ should respond with error when calling callback(error) (79ms)
      ✓ should respond with only the last action when there are more than one in the serverless.yml file (61ms)
      ✓ should not send to non existing client (50ms)
      ✓ should connect & disconnect (187ms)
      ✓ should receive correct call info (event only) (91ms)
      ✓ should be able to parse query string in connect (107ms)
      ✓ should be able to get headers in connect (167ms)
      ✓ should be able to get Authorization header in connect (78ms)
      ✓ should be able to receive messages via REST API (78ms)
      ✓ should receive error code when sending to a recently closed client via REST API (61ms)
      ✓ should be able to close connections via REST API (50ms)
      ✓ should receive error code when deleting a previously closed client via REST API (85ms)


  35 passing (2s)
```
In order to run the tests you should do in `.`:

- `npm i`.

- `npm link`.

Then in `./src/__tests__/manual/websocket':

- `npm run deploy-dynamodb-local`. Previously this command didn't block but now if it blocks then open another terminal window.

- Set you AWS credentials. I use `export AWS_PROFILE=XXXXX`.

- `npm run test-e2e`.

If you still have problems then report and I'll try to help.


> * merges from 'master' seem to be borked. not quite sure what to do about that. maybe we can revert the merges somehow, and redo them? [see below]

This is a big change for WebSocket. While this PR was alive, there were many changes in the code regarding WebSocket in `master`. The automatic merge stoped working a long time ago and I had to do manual merges. I'm sure to have missed things (even thought the test are green).


> * linting doesn't seem to work, that's probably on me though, I'll look into that.

Linting is working on my side. No linting => no commit :)


> * a bit of a problem is also: https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/hapi-plugin-websocket.js and https://github.com/computerpunc/serverless-offline/blob/websocket-fixes-authorizer/src/hapi-plugin-websocket.package.json
>   could you outline why you needed to pull the plugin in? what did you add, or change, and for what reason? is that something what could have also been added in a PR for `hapi-plugin-websocket`?

The 'hapi-plugin-websocket` is faulted. It doesn't really follow WebSocket protocol specifically the HTTP handshake part ($connect). And since no one is really maintaining that plugin, I therefore decided to pull it in and make the required changes.

> I'm thinking of pulling in the PR into a branch, same as we did with `websockets`, if that's alright with you?

Sure.

dnalborczyk commented:
branch: https://github.com/dherault/serverless-offline/tree/websocket-fixes-authorizer

dnalborczyk commented:
@computerpunc  I tried to merge/cherry-pick from the PR/branch. it's impossible.

I'm gonna move the tests around in master, and cherry-pick your tests on top of it in the `auth` branch.
then I try a merge with the remaining files (which is "only" Serverless.Offline, APIGatewayWebSocket and websocketHelper I believe).

computerpunc commented:
@dnalborczyk 
> I tried to merge/cherry-pick from the PR/branch. it's impossible.

 I totally understand. I had hard time too unfortunately. 

> I'm gonna move the tests around in master, and cherry-pick your tests on top of it in the `auth` branch.

OK. Where exactly are you going to move the tests around. `auth` to `master` or also in a different location in `master`?

> then I try a merge with the remaining files (which is "only" Serverless.Offline, APIGatewayWebSocket and websocketHelper I believe).

Yes, these are the files. If cherry-picking from `auth` to `master` didn’t work, you can try the other way around. Push these files as is to `master` (you basically overwrite) and then cherry-pick previous changes from `master`

dnalborczyk commented:
@computerpunc  I'm still planning on pulling this in! Mind if I re-open?

computerpunc commented:
> @computerpunc I'm still planning on pulling this in! Mind if I re-open?

@dnalborczyk 

Pulling-this-in in the way that GitHub intended a PR to be is no longer possible. There's not a shred of similarity left between `master` and this PR. 

Of course, the ideas in the PR still have merit. It can be seen as some sort of a prototype. That's why I haven't deleted the branch of this PR so anyone who needs can cherry pick any available code or test.

So from my perspective, there's no longer a need to keep this PR open. If it helps you keeping it open in anyway then sure, you can reopen.
Same goes for #778.

dnalborczyk commented:
> Pulling-this-in in the way that GitHub intended a PR to be is no longer possible. There's not a shred of similarity left between `master` and this PR.

yeah, I guess I might have to copy/paste it in. We'll see. I definitely don't want to discard any of this too easily, as I'm sure if took a while to implement and to test.

> Of course, the ideas in the PR still have merit. 

👍

dnalborczyk commented:
@computerpunc 

I haven't looked too deeply yet, is the main functionality in this method:
https://github.com/dherault/serverless-offline/pull/732/files#diff-9dc6f987be70c5ff99fd93f51a757739R372 

and https://github.com/dherault/serverless-offline/pull/732/files#diff-9dc6f987be70c5ff99fd93f51a757739R128

as well as: https://github.com/dherault/serverless-offline/pull/732/files#diff-d6eed065f710956595e85a590d36f253R247 ?

what is `createFunctionWithNoEvent` being used for? as I haven't used api gateway authorizer, is there any functionality which could potentially be shared with (or already is)?  I assume it's this module: https://github.com/dherault/serverless-offline/blob/master/src/api-gateway/createAuthScheme.js

computerpunc commented:
> I haven't looked too deeply yet, is the main functionality in this method:
> https://github.com/dherault/serverless-offline/pull/732/files#diff-9dc6f987be70c5ff99fd93f51a757739R372
> 
> and https://github.com/dherault/serverless-offline/pull/732/files#diff-9dc6f987be70c5ff99fd93f51a757739R128
> 
> as well as: https://github.com/dherault/serverless-offline/pull/732/files#diff-d6eed065f710956595e85a590d36f253R247 ?

Yes this is it.

> what is `createFunctionWithNoEvent` being used for? as I haven't used api gateway authorizer, is there any functionality which could potentially be shared with (or already is)? I assume it's this module: https://github.com/dherault/serverless-offline/blob/master/src/api-gateway/createAuthScheme.js

`createFunctionWithNoEvent` is just a place where authorizer method handler is created and later to be picked up by `createConnectWithAutherizerAction`.
`createAuthScheme` is more like all of the above code together. 
Take into account that the default WebSocket authorization header in `serverless` is `Auth` (a bad decision by the person who implemented authorizer support in WebSocket `serverless` as the default in HTTP is `Authorization`).

dnalborczyk commented:
> `createFunctionWithNoEvent` is just a place where authorizer method handler is created and later to be picked up by `createConnectWithAutherizerAction`.

I think this also gives me a better idea on how to better distinguish between the `handler function` and the events (http, websocket, and potentially eventually others).

> `createAuthScheme` is more like all of the above code together.
> Take into account that the default WebSocket authorization header in `serverless` is `Auth` (a bad decision by the person who implemented authorizer support in WebSocket `serverless` as the default in HTTP is `Authorization`).

ah, thanks for pointing that out! Not sure if it's a mistake, or was done on purpose. maybe it's a 1:1 mapping between `serverless` and `aws cloudformation`? if not, `serverless` might be willing to fix this if someone would make them aware of this.

Gr8z commented:
+1 would love to test Authorizers locally.

592da commented:
any progress guys ?
thanks for your help :)

ricardovf commented:
Will this be merged?

computerpunc commented:
> Will this be merged?

This PR cannot be merged per se. The current `master` and this PR do not have much in common anymore (if anything at all) as the `master` has been heavily refactored almost to the level of being written from scratch. The only reason this PR is still open is as a reminder to @dnalborczyk to add it in the future.

Since I wrote this PR, browser WebSocket authorization has moved from 'Authorization' header to query string, i.e. it is no longer possible to send an 'Authorization' header via the browser's WebSocket API and sending a token via query string is the only remaining option.
So `identitySource`  `route.request.querystring...` is more important than `route.request.header...` (if only one is supported and not both options).
